### PR TITLE
File upload

### PR DESCRIPTION
**Pull request by gauravmakkar**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -476,6 +476,119 @@ class ConnectionParameters {
...
+    soapCall.writeDocument("entity", DomUtil.parse(DomUtil.toXMLString(fileRes)));
+    await client._makeSoapCall(soapCall)
+  }
+
+  /**
+   * Will return the public URL of the uploaded file.
+   * @param fileRes
+   * @returns {Promise<string|*>}
+   */
+  const getPublicUrl = async (fileRes) => {
```
Comment (blfakir) [inline src/client.js:533]: Please use the the xtkFileRes object as in the comment before

File: src/client.js [hunk 2]
```diff
@@ -476,6 +476,119 @@ class ConnectionParameters {
...
+    upload: async (file) => {
+      return new Promise(async (resolve, reject) => {
+        if (!Util.isBrowser()) {
+          reject('File uploading is only supported in browser based calls.');
+        }
+        try {
+          var data = new FormData()
+          data.append('file_noMd5', file)
+          //TODO: Needs to be refactored after cookie issue get resolved.
+          const response = await fetch(`${client._connectionParameters._endpoint}/nl/jsp/uploadFile.jsp`, {
```
Comment (blfakir) [inline src/client.js:550]: Did you tried using **request** defined in  https://github.com/adobe/acc-js-sdk/blob/master/src/transport.js

File: src/client.js [hunk 3]
```diff
@@ -476,6 +476,119 @@ class ConnectionParameters {
...
+    soapCall.writeDocument("entity", DomUtil.parse(DomUtil.toXMLString(fileRes)));
+    await client._makeSoapCall(soapCall)
+    soapCall.getNextElement();
+    return soapCall.getNextString();
+  }
+  return {
+    upload: async (file) => {
+      return new Promise(async (resolve, reject) => {
+        if (!Util.isBrowser()) {
+          reject('File uploading is only supported in browser based calls.');
```
Comment (blfakir) [inline src/client.js:544]: Please add a test for this line

File: src/client.js [hunk 4]
```diff
@@ -476,6 +476,119 @@ class ConnectionParameters {
...
+// File Uploader
+// ========================================================================================
+
+const fileUploader = (client) => {
+  /**
+   * Will increament the counter
+   * @returns {Promise<number|*>}
+   */
+  const increaseValue = async () => {
+    const soapCall = client._prepareSoapCall('xtk:counter', 'IncreaseValue')
```
[the remark this answers is not in the document]
Comment (gauravmakkar) [reply src/client.js:489]: Other APIs in **client.js** are also using **_prepareSoapCall** for making soap calls. Is there any reason to use xtkSession instead of _prepareSoapCall?

File: README.md [hunk 5]
```diff
@@ -1452,7 +1452,21 @@ var event = await query.executeQuery();
 console.log(`>> Event: ${JSON.stringify(event)}`);
 ```
 
-
+# File Upload
```
Comment (mkiki) [inline README.md:1455]: As discussed previously, can you indicate the limitation, i.e. in which conditions the API works (for instance, that it is only supported when used in the browser)

File: README.md [hunk 6]
```diff
@@ -1452,7 +1452,21 @@ var event = await query.executeQuery();
...
+This API is used to upload a file and return the metadata that needs to be saved in delivery.
+```
+await client.fileUploader.upload(file);
```
Comment (mkiki) [inline README.md:1458]: Explain what file is. Which type, etc.

File: src/client.js [hunk 7]
```diff
@@ -10,14 +10,14 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 (function() {
-"use strict";
```
Comment (mkiki) [inline src/client.js:13]: Lots of changes in blank spaces is making the diff confusing. If you want to do some cleanup, can you do a separate PR?
In fact, it would be good to have a separate PR which removes whitespaces and add linting rules to make sure this remains correct

File: src/client.js [hunk 8]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+// ========================================================================================
+// File Uploader
+// ========================================================================================
+
+const fileUploader = (client) => {
+  /**
+   * Will increament the counter
+   * @returns {Promise<number|*>}
+   */
+  const increaseValue = async () => {
```
Comment (mkiki) [inline src/client.js:488]: Private functions should have “_” prefix (increaseValue, etc.)
If not, they are considered public, and you'll have to maintain their signature and compatibility forever

File: src/client.js [hunk 9]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+      const xtkCounter= await client.NLWS.xtkCounter.increaseValue({name: 'xtkResource'});
+      return xtkCounter
+  }
+
+  const createFileRes = (counter, data) => {
+      return {
+          internalName: 'RES' + counter,
+          md5: data[0].md5,
```
Comment (mkiki) [inline src/client.js:496]: why data[0] ? what if the array is empty? has more than 1 element?
This needs a short comment explaining the intent and the rationale of passing an array of 1 element.

File: src/client.js [hunk 10]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+  const publishIfNeeded = async (fileRes) => {
+      await client.NLWS.xtkFileRes.create(fileRes).publishIfNeeded();
+  }
+
+  /**
+   * Will return the public URL of the uploaded file.
+   * @param fileRes
+   * @returns {Promise<string|*>}
+   */
+  const getPublicUrl = async (fileRes) => {
```
Comment (mkiki) [inline src/client.js:530]: During previous discussion, we said that this should be returned by the upload call directly. Do we still need this function?

File: src/client.js [hunk 11]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+  return {
+    upload: async (file) => {
+      return new Promise(async (resolve, reject) => {
+        if (!Util.isBrowser()) {
+          reject('File uploading is only supported in browser based calls.');
+        }
+        try {
+          var data = new FormData()
+          data.append('file_noMd5', file)
+          //TODO: Needs to be refactored after cookie issue get resolved.
```
Comment (mkiki) [inline src/client.js:542]: This is not really clear. If the code does not work, I'm not sure we should commit it.

File: src/client.js [hunk 12]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+          const response = await fetch(`${client._connectionParameters._endpoint}/nl/jsp/uploadFile.jsp`, {
```
Comment (mkiki) [inline src/client.js:543]: You should use the "transport" interface, not fetch directly, i.e. client._transport(...)

File: src/client.js [hunk 13]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+            credentials: 'include',
+            method: 'POST',
+            body: data,
+            headers: {
+              'x-security-token': client._securityToken,
+              'Cookie': '__sessiontoken=' + client._sessionToken,
+            }
+          })
+          const okay = await response.text()
+          var iframe = document.createElement('iframe');
```
Comment (mkiki) [inline src/client.js:554]: Use const

File: src/client.js [hunk 14]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+          iframe.style.height = 0;
+          iframe.style.width = 0;
+          document.controller = {
```
Comment (mkiki) [inline src/client.js:557]: This requires a comment. There are a lot of assumptions behind those lines of code

File: src/client.js [hunk 15]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+            uploadFileCallBack: async (data) => {
+              const counter = await increaseValue(); // Step 1
+              const fileRes= createFileRes(counter, data)
```
Comment (mkiki) [inline src/client.js:560]: missing space before the '=' symbol

File: src/client.js [hunk 16]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+              await write(fileRes); // Step 2
+              await publishIfNeeded(fileRes) // Step 3
+              const url = await getPublicUrl(fileRes) // Step 3
+              resolve({
+                label: data[0].fileName,
```
Comment (mkiki) [inline src/client.js:565]: Again, it is not clear why data is an array of one element. Maybe more comments would help understand

File: src/client.js [hunk 17]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+                md5: data[0].md5,
+                type: file.type,
+                size: file.size,
+                url: url
+              })
+            }
+          }
+          var html = `<body>${okay}</body>`;
```
Comment (mkiki) [inline src/client.js:573]: Can you guarantee that the lack of escaping does not cause any security issue?

File: src/client.js [hunk 18]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+          document.body.appendChild(iframe);
+          iframe.contentWindow.document.open();
+          iframe.contentWindow.document.write(html);
+          iframe.contentWindow.document.close();
+        } catch (ex) {
+          reject(ex);
```
Comment (mkiki) [inline src/client.js:579]: Error management is not consistent with the rest of the SDK and needs to be fixed

File: src/client.js [hunk 19]
```diff
@@ -526,12 +633,18 @@ class Client {
 
         // expose utilities
 
-        /** 
+        /**
+        * File Uploader API
+        * @type {{upload: (function(*=): Promise<unknown>)}}
```
Comment (mkiki) [inline src/client.js:638]: Type definition should be more accurate

File: src/client.js [hunk 20]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+  /**
+   * Will return the public URL of the uploaded file.
+   * @param fileRes
+   * @returns {Promise<string|*>}
+   */
+  const getPublicUrl = async (fileRes) => {
+      return await client.NLWS.xtkFileRes.create(fileRes).getURL()
+  }
+  return {
+    upload: async (file) => {
```
Comment (mkiki) [inline src/client.js:534]: Upload function should be documented. Make sure to generate the JSDoc and that the generated doc is complete and acurate

File: src/client.js (hunk 17 above)
Comment (mkiki) [inline src/client.js:573]: Use const

File: test/client.test.js [hunk 21]
```diff
@@ -2803,4 +2803,12 @@ describe('ACC Client', function () {
             });
         });
     });
+
```
Comment (mkiki) [inline test/client.test.js:2806]: This is far from enough from testing. We want 100% coverage

File: src/client.js [hunk 22]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+          md5: data[0].md5,
+          label: data[0].fileName,
+          fileName: data[0].fileName,
+          originalName: data[0].fileName,
+          useMd5AsFilename: '1',
+          storageType: 5,
+          xtkschema: 'xtk:fileRes'
+
+      }
+  }
```
Comment (mkiki) [inline src/client.js:505]: Indentation seems to be incorrect (it's hard to say in github merge view)

File: src/client.js (hunk 16 above)
Comment (mkiki) [inline src/client.js:565]: Why is the attribute containing the file name called "label"?

File: README.md [hunk 23]
```diff
@@ -1452,7 +1452,21 @@ var event = await query.executeQuery();
...
+```
+will return
+```
+{
+  label: <file-name>,
+  md5: <md5-of-file-content>,
+  type: <file-type>,
```
Comment (mkiki) [inline README.md:1465]: what are the possible values for type? Is it the extension? 
Make sure there's a comment so that it is not ambiguous

**Review by mkiki — CHANGES_REQUESTED** [review]
Hi @gauravmakkar , please see comments. The code seems to work, but there are a lot of small details to polish and finish before it can be merged. In particular with regards to tests and documentation

File: src/client.js (hunk 10 above)
Comment (gauravmakkar) [reply src/client.js:530]: This is the private function getting called inside of upload function.

File: src/client.js [hunk 24]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
-// ========================================================================================
-// ACC Client
-// ========================================================================================
+  // ========================================================================================
+  // File Uploader
+  // ========================================================================================
+
+  const fileUploader = (client) => {
+    /**
+         * Will call the SOAP method(IncreaseValue) to increament the counter
```
Comment (mkiki) [inline src/client.js:502]: Please fix indentation before commit

Comment (mkiki) [inline src/client.js:502]: I'm not too fan of those comments which do not really provide any information. Can you be a little more explicit: what is this counter about? Imagine someone reading this code who is not deeply familiar with Campaign internals. I think the comment should give some hints about why we need to increase a counter, and what this counter is about.

File: src/client.js [hunk 25]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+    const _increaseValue = async () => {
+      const xtkCounter= await client.NLWS.xtkCounter.increaseValue({name: 'xtkResource'});
+      return xtkCounter
+    }
+
+    /**
+         * This will create and return fileRes object
+         * @param counter
+         * @param data
+         * @returns {{originalName, internalName: string, fileName, useMd5AsFilename: string, xtkschema: string, storageType: number, label, md5: (string|string|string|null|any|number)}}
```
Comment (mkiki) [inline src/client.js:514]: Are you sure? Looks like some types are missing, and the md5 type is weird

File: src/client.js [hunk 26]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+  }
 
-// ========================================================================================
-// ACC Client
-// ========================================================================================
+  // ========================================================================================
+  // File Uploader
+  // ========================================================================================
+
+  const fileUploader = (client) => {
```
Comment (mkiki) [inline src/client.js:500]: JSDoc annotations are missing and the function appears in a global section. I think it should be in the Campaign namespace. You can use npm run jsdoc to generate the jsdoc.

Comment (mkiki) [inline src/client.js:500]: Is there a particular reason not to use a class as this is done everywhere else in the SDK?

File: src/client.js [hunk 27]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+        fileName: data[0].fileName,
+        originalName: data[0].fileName,
+        useMd5AsFilename: '1',
+        storageType: 5,
+        xtkschema: 'xtk:fileRes'
+
+      }
+    }
+
+    /**
```
Comment (mkiki) [inline src/client.js:531]: This comment is not really helpful. Is there a particular reason to have created separate functions for each of the API calls? I mean, what's the point of calling this._write(fileRes) vs. simply client.NLWS.xtkSession.write(fileRes) ? It seems adding a function here makes things most verbose and does not really help understand the intent of the code

File: src/client.js [hunk 28]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+    return {
+      /**
+             * This is the exposed/public method for fileUploader instance which will do all the processing related to the upload process internally and returns the promise containing all the required data.
+             * @param file
+             * @returns {Promise<{name: string, md5: string, type: string, size: string, url: string}>}
+             */
+      upload: (file) => {
+        return new Promise((resolve, reject) => {
+          if (!Util.isBrowser()) {
+            reject('File uploading is only supported in browser based calls.');
```
Comment (mkiki) [inline src/client.js:569]: All exception types must be consistent (see CampaignException)

File: src/client.js [hunk 29]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+         * @returns {Promise<string>}
+         * @private
+         */
+    const _getPublicUrl = async (fileRes) => {
+      return await client.NLWS.xtkFileRes.create(fileRes).getURL()
+    }
+    return {
+      /**
+             * This is the exposed/public method for fileUploader instance which will do all the processing related to the upload process internally and returns the promise containing all the required data.
+             * @param file
```
Comment (mkiki) [inline src/client.js:563]: Can you describe what file is? (even if it's redundant with the readme)

File: src/client.js [hunk 30]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+                  })
+                }
+              }
+              var html = `<body>${okay}</body>`;
+              document.body.appendChild(iframe);
+              iframe.contentWindow.document.open();
+              iframe.contentWindow.document.write(html);
+              iframe.contentWindow.document.close();
+            }).catch((ex)=>{
+              reject(ex);
```
Comment (mkiki) [inline src/client.js:611]: All exception types must be consistent (see CampaignException)
The idea is to create an error code and throw a CampaignException with the proper error code (and reuse error codes when possible or when it makes sense)

File: src/client.js [hunk 31]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+            })
+          } catch (ex) {
+            reject(ex);
```
Comment (mkiki) [inline src/client.js:614]: All exception types must be consistent (see CampaignException)

File: src/client.js [hunk 32]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+             * @param file
+             * @returns {Promise<{name: string, md5: string, type: string, size: string, url: string}>}
+             */
+      upload: (file) => {
+        return new Promise((resolve, reject) => {
+          if (!Util.isBrowser()) {
+            reject('File uploading is only supported in browser based calls.');
+          }
+          try {
+            var data = new FormData()
```
Comment (mkiki) [inline src/client.js:572]: const

File: src/client.js [hunk 33]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+              processData: false,
+              credentials: 'include',
+              method: 'POST',
+              body: data,
+              headers: {
+                'x-security-token': client._securityToken,
+                'Cookie': '__sessiontoken=' + client._sessionToken,
+              }
+            }).then((okay)=>{
+              var iframe = document.createElement('iframe');
```
Comment (mkiki) [inline src/client.js:586]: const

File: src/client.js [hunk 34]
```diff
@@ -305,427 +305,558 @@ class ConnectionParameters {
...
+                  resolve({
+                    name: data[0].fileName,
+                    md5: data[0].md5,
+                    type: file.type,
+                    size: file.size,
+                    url: url
+                  })
+                }
+              }
+              var html = `<body>${okay}</body>`;
```
Comment (mkiki) [inline src/client.js:605]: const

File: test/client.test.js [hunk 35]
```diff
@@ -2502,387 +2504,546 @@ describe('ACC Client', function () {
...
+        type: "text/html",
+        url: "http://hello.com"
+      });
+    })
+
+    it('throws error with dependant failures', async () => {
+      // Create a mock client and logon
+      const client = await Mock.makeClient();
+      client._transport.mockReturnValueOnce(Mock.LOGON_RESPONSE);
+      client._transport._xxx = 3;
```
Comment (mkiki) [inline test/client.test.js:3015]: Remove this

File: test/mock.js [hunk 36]
```diff
@@ -614,6 +614,523 @@ const GET_HELLO_RESPONSE = Promise.resolve(`<?xml version='1.0'?>
...
+            <plValue xsi:type='xsd:int'>1</plValue>
+        </IncreaseValueResponse>
+    </SOAP-ENV:Body>
+</SOAP-ENV:Envelope>
+
+`);
+
+const FILE_RES_WRITE_RESPONSE = Promise.resolve(`<?xml version='1.0'?><SOAP-ENV:Envelope xmlns:xsd='http://www.w3.org/2001/XMLSchema' xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance' xmlns:ns='urn:wpp:default' xmlns:SOAP-ENV='http://schemas.xmlsoap.org/soap/envelope/'><SOAP-ENV:Body><GetEntityIfMoreRecentResponse xmlns='urn:wpp:default' SOAP-ENV:encodingStyle='http://schemas.xmlsoap.org/soap/encoding/'><pdomDoc xsi:type='ns:Element' SOAP-ENV:encodingStyle='http://xml.apache.org/xml-soap/literalxml'><schema _cs="Sessions to the application server (xtk)" created="2022-02-03 12:55:02.621Z" createdBy-id="0" dependingSchemas="" desc="Sessions to the application server" entitySchema="xtk:schema" img="" implements="xtk:persist" label="Sessions to the application server" labelSingular="Session" lastModified="2022-07-11 02:29:30.026Z" library="true" mappingType="xmlFile" md5="9D716FE59E174E87C5BA8A834DD4B11B" modifiedBy-id="0" name="session" namespace="xtk" xtkschema="xtk:schema">
```
Comment (mkiki) [inline test/mock.js:676]: Can you narrow it down to just what is needed like in GET_FILERES_QUERY_SCHEMA_RESPONSE?
Having very large payloads like this is not practical for unit tests.
Also, this seems to be the payload for a method returning a schema, definitely not a call to writing a file res. Either your payload is wrong or the constant name is wrong, or both, or maybe there's an issue in the test itself

**Review by mkiki — CHANGES_REQUESTED** [review]
Hi Gaurav, we're almost there. The code looks good but there's still some polishing to do before it can be merged. In particular in the area of error management which needs to be consistent with the rest

File: src/client.js (hunk 26 above)
Comment (gauravmakkar) [reply src/client.js:500]: SDK follows both of the approaches. 
For example:
`const clientHandler = (representation, headers, pushDownOptions) => {` 
and 
`class ConnectionParameters {`

i just picked one which suits better for this use case.

File: src/client.js (hunk 27 above)
Comment (gauravmakkar) [reply src/client.js:531]: It was done to consider the separation of concerns.

File: src/client.js (hunk 26 above)
Comment (gauravmakkar) [reply src/client.js:500]: Tried running this task, it didn't change anything.

File: src/client.js [hunk 37]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
 
 }
 
+// ========================================================================================
+// File Uploader
+// ========================================================================================
+
+const fileUploader = (client) => {
+  /**
```
[the remark this answers is not in the document]
Comment (gauravmakkar) [reply src/client.js:500]: Done

File: src/client.js (hunk 26 above)
Comment (gauravmakkar) [reply src/client.js:505]: Added additional jsdocs

File: src/client.js (hunk 28 above)
Comment (gauravmakkar) [reply src/client.js:577]: Done

File: src/client.js [hunk 38]
```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
...
+   * @returns {Promise<string|*>}
+   */
+  const getPublicUrl = async (fileRes) => {
+      return await client.NLWS.xtkFileRes.create(fileRes).getURL()
+  }
+  return {
+    upload: async (file) => {
+      return new Promise(async (resolve, reject) => {
+        if (!Util.isBrowser()) {
+          reject('File uploading is only supported in browser based calls.');
```
[the remark this answers is not in the document]
Comment (gauravmakkar) [reply src/client.js]: Done

File: src/client.js (hunk 13 above)
Comment (gauravmakkar) [reply src/client.js]: Done

File: src/client.js (hunk 15 above)
Comment (gauravmakkar) [reply src/client.js]: Done

File: src/client.js (hunk 14 above)
Comment (gauravmakkar) [reply src/client.js]: Done

File: src/client.js (hunk 16 above)
Comment (gauravmakkar) [reply src/client.js]: Done

File: src/client.js (hunk 18 above)
Comment (gauravmakkar) [reply src/client.js]: Done

File: src/client.js (hunk 19 above)
Comment (gauravmakkar) [reply src/client.js]: done

File: README.md [hunk 39]
```diff
@@ -1504,7 +1504,21 @@ var event = await query.executeQuery();
 console.log(`>> Event: ${JSON.stringify(event)}`);
 ```
 
-
+# File Upload
+This API is used to upload a file and return the metadata that needs to be saved in delivery. Currently, this API is supported only in Browsers.
```
Comment (mkiki) [inline README.md:1508]: * return => returns (typo)
* upload is a generic mechanism. It _can_ be used in deliveries, but is not limited to deliveries. Your comment seems to imply that it does.
* Browsers => browsers (no need for upper case)

File: README.md [hunk 40]
```diff
@@ -1504,7 +1504,21 @@ var event = await query.executeQuery();
 console.log(`>> Event: ${JSON.stringify(event)}`);
 ```
 
-
+# File Upload
+This API is used to upload a file and return the metadata that needs to be saved in delivery. Currently, this API is supported only in Browsers. 
+```
+await client.fileUploader.upload(file);
```
Comment (mkiki) [inline README.md:1510]: I would write 

const metadata = await client.fileUploader.upload(file);

File: README.md [hunk 41]
```diff
@@ -1504,7 +1504,21 @@ var event = await query.executeQuery();
...
-
+# File Upload
+This API is used to upload a file and return the metadata that needs to be saved in delivery. Currently, this API is supported only in Browsers. 
+```
+await client.fileUploader.upload(file); 
+```
+where file is an instance of [File](https://developer.mozilla.org/en-US/docs/Web/API/File), will return
+```
+{
+  name: <name-referenceed-by-file-object>, // file.name
```
Comment (mkiki) [inline README.md:1515]: Check for typos: referenceed => referenced
What is the purpose of comments such as "file.name"? They do not seem to bring much value

File: src/campaign.js [hunk 42]
```diff
@@ -38,7 +38,8 @@ const { Util } = require("./util.js");
   static NOT_LOGGED_IN(call, details)                     { return new CampaignException(     call, 400, 16384, `SDK-000010 Cannot call API because client is not logged in`, details); }
   static DECRYPT_ERROR(details)                           { return new CampaignException(undefined, 400, 16384, `SDK-000011 "Cannot decrypt password: password marker is missing`, details); }
   static SESSION_EXPIRED()                                { return new CampaignException(undefined, 401, 16384, `SDK-000012 "Session has expired or is invalid. Please reconnect.`); }
-  
+  static FILE_UPLOAD_FAILED(details)                      { return new CampaignException(undefined, 500, 16384, `SDK-000013 "Unknown error while uploading file`, details); }
```
Comment (mkiki) [inline src/campaign.js:41]: I'm not sure it's correct to say "Unknown error". Error details will be provided in nested exceptions. I think the message should just say "Failed to upload file ..." and should mention the file name for better troubleshooting

File: src/client.js (hunk 26 above)
Comment (mkiki) [reply src/client.js:505]: Good point. clientHandler should also use classes, I'll change it.
Can you change to use class as well?
Maybe I'm missing something, but I'm not sure I understand why using a function is better here. It seems to me it's harder to read

File: src/client.js (hunk 27 above)
Comment (mkiki) [reply src/client.js:539]: What do you mean?

File: src/client.js (hunk 28 above)
Comment (mkiki) [reply src/client.js:577]: I don't believe this will return the right kind of exception: calling reject here will not be caught by the catch blocks below

File: src/client.js [hunk 43]
```diff
@@ -305,427 +305,574 @@ class ConnectionParameters {
...
+              url: `${client._connectionParameters._endpoint}/nl/jsp/uploadFile.jsp`,
+              processData: false,
+              credentials: 'include',
+              method: 'POST',
+              body: data,
+              headers: {
+                'x-security-token': client._securityToken,
+                'Cookie': '__sessiontoken=' + client._sessionToken,
+              }
+            }).then((okay)=>{
```
Comment (mkiki) [inline src/client.js:593]: indentation is not correct. There are other occurrences as well later in the code

File: src/client.js [hunk 44]
```diff
@@ -305,427 +305,574 @@ class ConnectionParameters {
...
+              }
+              const iframe = document.createElement('iframe');
+              iframe.style.height = 0;
+              iframe.style.width = 0;
+              document.controller = { // Written to  support https://git.corp.adobe.com/Campaign/ac/blob/v6-master/nl/datakit/nl/jsp/uploadFile.jsp
+                uploadFileCallBack: async (data) => {
+                  if(!data || data.length !== 1){
+                    // Tried to replicate the logic for file upload functionality written here:
+                    // https://git.corp.adobe.com/Campaign/ac/blob/v6-master/wpp/xtk/web/dce/uploader.js
+                    reject(CampaignException.FILE_UPLOAD_FAILED('MALFORMED DATA'+ data.toString()));
```
Comment (mkiki) [inline src/client.js:605]: This is not proper english. Logging potentially very long variables (data) should be avoided

File: src/client.js [hunk 45]
```diff
@@ -305,427 +305,574 @@ class ConnectionParameters {
...
+                  }
+                  const counter = await _increaseValue(); // Step 1
+                  const fileRes = _createFileRes(counter, data)
+                  await _write(fileRes); // Step 2
+                  await _publishIfNeeded(fileRes) // Step 3
+                  const url = await _getPublicUrl(fileRes) // Step 3
```
Comment (mkiki) [inline src/client.js:611]: Shouldn't this be Step 4?

**Review by mkiki — CHANGES_REQUESTED** [review]
Hi Gaurav, there are still things missing (details) or previous remarks that I think should be considered. Also, I do not think we should publish this function until we have the server side changes ready because this code in this PR will not work unless someone manually sets cookies. Do you know if they are ready yet or when they will be?

**Review by mkiki — CHANGES_REQUESTED** [review]
Unit tests are failing. Can you make sure all tests pass, are 100% and that all bullet point items described in the "# Build & Run" section of the readme are also all good?

**Review by mkiki — APPROVED** [review]
